### PR TITLE
WIP: Add monitor non interactive mode and fix tunnel proxy core functionality 

### PR DESCRIPTION
**Pull request by t-mon**

### CODE REVIEW

File: libnymea-remoteproxy/server/tcpsocketserver.cpp [hunk 1]
```diff
@@ -175,32 +191,50 @@ void SslServer::incomingConnection(qintptr socketDescriptor)
...
+    case QAbstractSocket::SocketTimeoutError:
+    case QAbstractSocket::DatagramTooLargeError:
+    case QAbstractSocket::NetworkError:
+    case QAbstractSocket::SslHandshakeFailedError:
+    case QAbstractSocket::UnfinishedSocketOperationError:
+    case QAbstractSocket::SslInternalError:
+    case QAbstractSocket::SslInvalidUserDataError:
+    case QAbstractSocket::TemporaryError:
+    case QAbstractSocket::UnknownSocketError:
+        qCWarning(dcTcpSocketServer()) << "Explitily closing the socket due to error.";
```
Comment (mzanetti) [inline libnymea-remoteproxy/server/tcpsocketserver.cpp:231]: typo

File: libnymea-remoteproxy/server/tcpsocketserver.cpp [hunk 2]
```diff
@@ -175,32 +191,50 @@ void SslServer::incomingConnection(qintptr socketDescriptor)
...
+        sslSocket->close();
+        break;
+    default:
+        break;
```
Comment (mzanetti) [inline libnymea-remoteproxy/server/tcpsocketserver.cpp:235]: Is there a reason why it's only closed for the above errors and not for the others?

File: libnymea-remoteproxy/tunnelproxy/tunnelproxyserver.cpp [hunk 3]
```diff
@@ -370,13 +395,16 @@ void TunnelProxyServer::onClientDataAvailable(const QUuid &clientId, const QByte
                     TunnelProxyClientConnection *clientConnection = serverConnection->getClientConnection(frame.socketAddress);
                     if (!clientConnection) {
                         qCWarning(dcTunnelProxyServer()) << "The server connection wants to send data to a client connection which has not been registered to the server.";
-                        // FIXME: tell the server this client does not exist
-                        return;
+                        qCWarning(dcTunnelProxyServer()) << "Notifying the server that there is no longer any socket connected with address" << frame.socketAddress;
+                        QVariantMap params;
+                        params.insert("socketAddress", frame.socketAddress);
+                        m_jsonRpcServer->sendNotification("TunnelProxy", "ClientDisconnected", params, serverConnection->transportClient());
```
Comment (mzanetti) [inline libnymea-remoteproxy/tunnelproxy/tunnelproxyserver.cpp:401]: If the server sends something to a client which has long gone, IMO the problem happened earlier. Why did the server not get informed about this directly when the client went away?

Comment (mzanetti) [reply libnymea-remoteproxy/tunnelproxy/tunnelproxyserver.cpp:401]: I guess in theory there could be some race condition, where the server is informed but already has put a message for a particular client to the output buffer. In that case this additional ClientDisconnected message seems unneeded too.

File: nymea-remoteproxy.conf [hunk 4]
```diff
@@ -15,7 +15,7 @@ authorizerLambdaFunction=system-services-authorizer-dev-checkToken
 awsCredentialsUrl=http://169.254.169.254/latest/meta-data/iam/security-credentials/EC2-Remote-Connection-Proxy-Role
 
 [SSL]
-enabled=true
+enabled=false
```
Comment (mzanetti) [inline nymea-remoteproxy.conf:18]: this seems like a unintentional addition

File: tunnelclient/clientconnection.cpp [hunk 5]
```diff
@@ -1,15 +1,22 @@
 #include "clientconnection.h"
 
-ClientConnection::ClientConnection(const QUrl &serverUrl, const QString &name, const QUuid &uuid, const QUuid &serverUuid, bool insecure, QObject *parent) :
+ClientConnection::ClientConnection(const QUrl &serverUrl, const QString &name, const QUuid &uuid, const QUuid &serverUuid, bool insecure, bool sendRandomData, QObject *parent) :
```
Comment (mzanetti) [inline tunnelclient/clientconnection.cpp:3]: IMO this is the wrong place... if you want a test to send random data, create a clientconnection and then write random data to it. There's no point for every client app having to pass false to this flag or even having the code for this.

File: libnymea-remoteproxy/engine.cpp [hunk 6]
```diff
@@ -284,16 +284,22 @@ void Engine::onTimerTick()
...
+            m_tunnelProxyServer->tick();
+        }
 
-        m_currentTimeCounter = 0;
+        //        QVariantMap serverStatistics = buildMonitorData();
+        //        m_monitorServer->updateClients(serverStatistics);
+
+        //        m_logEngine->logStatistics(serverStatistics.value("proxyStatistic").toMap().value("tunnelCount").toInt(),
+        //                                   serverStatistics.value("proxyStatistic").toMap().value("clientCount").toInt(),
+        //                                   serverStatistics.value("proxyStatistic").toMap().value("troughput").toInt());
```
[the remark this answers is not in the document]
Comment (t-mon) [reply libnymea-remoteproxy/engine.cpp:302]: meybe, not sure yet

File: libnymea-remoteproxy/server/tcpsocketserver.cpp (hunk 2 above)
Comment (t-mon) [reply libnymea-remoteproxy/server/tcpsocketserver.cpp:235]: the other errors seem to be a result of a disconnect or will disconnect the client, but I'm experimenting here, will try to close the socket on any error.

File: tunnelclient/clientconnection.cpp (hunk 5 above)
Comment (t-mon) [reply tunnelclient/clientconnection.cpp:3]: This is the client connection class of the cli, not of the lib. The client connection on it self (the cli) has no purpose without sending data besides testing the connection itself.
This way we have traffic in the tests, and there is no other reason for this cli tool.

**Comment by t-mon** [comment]
Deprecated, see #36

**Pull request closed without merging**